federation/schema.go

```diff
@@ -76,6 +76,52 @@ func validateFederationKeys(serviceNames []string, serviceSchemasByName map[stri
 	return nil
 }
 
+// validateFieldsReturningFederatedObject checks that if an object is returned by a field func it can not be a shadow object type
+func validateFieldsReturningFederatedObject(serviceNames []string, serviceSchemasByName map[string]*IntrospectionQueryResult, types map[string]graphql.Type, fieldInfos map[*graphql.Field]*FieldInfo) error {
+	for _, service := range serviceNames {
+		for _, typ := range serviceSchemasByName[service].Schema.Types {
+			if typ.Kind != "OBJECT" {
+				continue
+			}
+			for _, field := range typ.Fields {
+				// Check that the field's return type is an object
+				fieldReturnType := getRootType(field.Type)
+				if fieldReturnType.Kind != "OBJECT" {
+					continue
+				}
+				// Error if it is a shadow object. To check this
+				// (1) Look through all the fields on the object to see if there is a federation field (__federation) and that
+				// the federation field is not on the current service
+				// (2) Look through all the fields on the federation object to see if it has a field for <ObjectType>-<Service>
+				returnObj, ok := types[fieldReturnType.Name].(*graphql.Object)
+				if !ok {
+					return oops.Errorf("Return type %s is not a graphql object", fieldReturnType.Name)
+				}
+				for name, f := range returnObj.Fields {
+					if name == federationField && !fieldInfos[f].Services[service] {
+						federatedFieldName := fmt.Sprintf("%s-%s", fieldReturnType, service)
+						if field.Name != federatedFieldName {
```

Comment:
Why need this check?

Reply:
On the "Federation" object, we have a function that looks like  "objectName-serviceName" (eg User-s2). This is used to create the shadow object, but in this case it's allowed to return the "user" object type. We want to ignore this check then

Reply:
Maybe add this as a comment of this check and change it to 
`if field.Name == federatedFieldName {continue}` ?

Reply:
updated!